src/ReachSets/discretize.jl

```diff
@@ -1,310 +1,656 @@
+                                   sih_method::String="concrete")
+
+    if sih_method == "concrete"
+        sih = symmetric_interval_hull
+    elseif sih_method == "lazy"
+        sih = SymmetricIntervalHull
     else
-        if pade_expm
-            ϕ = padm(A*δ)
-        else
-            ϕ = expmat(Matrix(A*δ))
-        end
+        throw(ArgumentError("the method $sih_method is unknown"))
     end
 
-    # early return for homogeneous systems
-    if cont_sys isa IVP{<:LinearContinuousSystem}
-         Ω0 = CH(X0, ϕ * X0)
-        return DiscreteSystem(ϕ, Ω0)
-    end
-    U = inputset(cont_sys)
-    inputs = next_set(U, 1)
+    # unrwap coefficient matrix and initial states
+    A, X0 = 𝑆.s.A, 𝑆.x0
+
+    # compute matrix ϕ = exp(Aδ)
+    ϕ = exp_Aδ(A, δ, exp_method=exp_method)
 
     # compute the transformation matrix to bloat the initial states
-    if lazy_expm
-        P = SparseMatrixExp([abs.(A*δ) sparse(δ*I, n, n) spzeros(n, n);
-                             spzeros(n, 2*n) sparse(δ*I, n, n);
-                             spzeros(n, 3*n)])
-        Phi2Aabs = sparse(get_columns(P, (2*n+1):3*n)[1:n, :])
+    Phi2Aabs = Φ₂(abs.(A), δ, exp_method=exp_method)
+
+    if algorithm == "forward"
+        Einit = sih(Phi2Aabs * sih((A * A) * X0))     # use Eplus
+    elseif algorithm == "backward"
+        Einit = sih(Phi2Aabs * sih((A * A * ϕ) * X0)) # use Eminus
     else
-        if pade_expm
-            P = padm([abs.(A*δ) sparse(δ*I, n, n) spzeros(n, n);
-                      spzeros(n, 2*n) sparse(δ*I, n, n);
-                      spzeros(n, 3*n)])
-        else
-            P = expmat(Matrix([abs.(A*δ) sparse(δ*I, n, n) spzeros(n, n);
-                           spzeros(n, 2*n) sparse(δ*I, n, n);
-                           spzeros(n, 3*n)]))
-        end
-        Phi2Aabs = P[1:n, (2*n+1):3*n]
+        throw(ArgumentError("the algorithm $approximation is unknown"))
     end
 
-    if isa(inputs, ZeroSet)
-        if approx_model == "forward" || approx_model == "backward"
-            Ω0 = CH(X0, ϕ * X0 + δ * inputs)
-        end
-    else
-        EPsi = sih(Phi2Aabs * sih(A * inputs))
-        discretized_U = δ * inputs + EPsi
-        if approx_model == "forward"
-            EOmegaPlus = sih(Phi2Aabs * sih((A * A) * X0))
-            Ω0 = CH(X0, ϕ * X0 + discretized_U + EOmegaPlus)
-        elseif approx_model == "backward"
-            EOmegaMinus = sih(Phi2Aabs * sih((A * A * ϕ) * X0))
-            Ω0 = CH(X0, ϕ * X0 + discretized_U + EOmegaMinus)
-        end
+    # early return for homogeneous systems
+    if inputdim(𝑆) == 0
+        Ω0 = ConvexHull(X0, ϕ * X0 ⊕ Einit)
+        return IVP(LDS(ϕ), Ω0)
     end
 
+    U = inputset(𝑆)
+    U0 = next_set(U, 1)
+
+    Eψ0 = sih(Phi2Aabs * sih(A * U0))
+    Ω0 = ConvexHull(X0, ϕ * X0 ⊕ δ*U0 ⊕ Eψ0 ⊕ Einit)
+
     if U isa ConstantInput
-        return DiscreteSystem(ϕ, Ω0, discretized_U)
+        Ud = map(ui -> δ*ui ⊕ Eψ0, U)
```

Comment:
Same comment as above: Why do you need the `map` here?

Reply:
I answered in the previous one; it is the preferred way to transform inputs in a type stable manner. In cases such as in `discretize_nobloating`, we can use map without having to consider the type of input.